mesonbuild/mesonmain.py

```diff
@@ -318,6 +337,7 @@ def run(original_args, mainfile=None):
 
     args = mesonlib.expand_arguments(args)
     options = parser.parse_args(args)
+    filter_meson_options(options, args)
```

Comment:
Nice

I thought maybe because this crosses with original_args, this could reintroduce something like #2811, but I think because `Environment.had_argument_for()`  already looks for -- and -D, that's going to be ok.